### PR TITLE
FEATURE: Add sorting to list

### PR DESCRIPTION
**Pull request by suffle**
This PR adds sorting to the list view. It reimplements nearly the same feature as in the old media browser.

Closes: #28

### CODE REVIEW

File: Classes/GraphQL/Resolver/Type/QueryResolver.php [hunk 1]
```diff
@@ -174,7 +179,17 @@ protected function createAssetProxyQuery(
             }
         }
 
-        // TODO: Implement sorting via `SupportsSortingInterface`
+        if ($sortBy && $assetProxyRepository instanceof SupportsSortingInterface) {
+            switch ($sortBy) {
```
Comment (suffle) [inline Classes/GraphQL/Resolver/Type/QueryResolver.php:183]: This implementation equals the one from the old media browser. I was thinking of a more flexible solution, e.g. go for `$assetProxyRepository->orderBy([$sortBy => $sortDirection]);` and leave the sortB completely to the frontend, but decided a more explicit version here would be better

File: Resources/Private/JavaScript/media-module/src/components/TopBar/TopBar.tsx [hunk 2]
```diff
@@ -24,7 +25,10 @@ const TopBar: React.FC = () => {
     const { selectionMode } = useMediaUi();
     const classes = useStyles({ selectionMode });
 
-    const components = useMemo(() => [ClipboardActions, SearchBox, TypeFilter, ViewModeSelector], []);
+    const components = useMemo(
+        () => [ClipboardActions, SearchBox, TypeFilter, SortOrderSelector, ViewModeSelector],
```
Comment (suffle) [inline Resources/Private/JavaScript/media-module/src/components/TopBar/TopBar.tsx:29]: Hopefully the positioning is OK here, I wasn't sure, if there are any designs / mockups on what goes where

File: Resources/Private/JavaScript/core/src/hooks/useAssetsQuery.ts [hunk 3]
```diff
@@ -14,7 +14,7 @@ import {
     initialLoadCompleteState,
 } from '../state';
 import { ASSETS } from '../queries';
-
+import selectedSortOrderState, { SORT_BY, SORT_DIRECTION } from '../state/selectedSortOrderState';
```
Comment (Sebobo) [inline Resources/Private/JavaScript/core/src/hooks/useAssetsQuery.ts:17]: You can combine the import into the line 14 above with the state variables.

File: Resources/Private/JavaScript/media-module/src/components/TopBar/TopBar.tsx (hunk 2 above)
Comment (Sebobo) [reply Resources/Private/JavaScript/media-module/src/components/TopBar/TopBar.tsx:29]: Yes that's fine, there are no official guidelines yet. I mostly used the ideas collected in the discuss topics about the media management.

File: Classes/GraphQL/Resolver/Type/QueryResolver.php (hunk 1 above)
Comment (Sebobo) [reply Classes/GraphQL/Resolver/Type/QueryResolver.php:183]: Yes that's better 👍

**Comment by Sebobo** [comment]
Thx ❤️
Will test this in the next days. Looks good so far by reading!

I would probably also like to have this persisted to keep the order when leaving the module and coming back later. But I encountered some issues with my initial implementation of persisted view options while implementing #115.
So this can be done later separately.

File: Resources/Private/JavaScript/core/src/hooks/useAssetsQuery.ts (hunk 3 above)
Comment (suffle) [reply Resources/Private/JavaScript/core/src/hooks/useAssetsQuery.ts:17]: So importing the selectedSortOrderState with line 14 and the enums separately? Oder do you propose exporting the enums through the state/index.ts as well? Or maybe there is an even better place where to put them?

**Comment by Sebobo** [comment]
@suffle the select box creates some issues when used in the UI in smaller screens:

Bildschirmfoto 2022-03-04 um 10 08 33

We should probably only show the icon when the space is limited and `selectionMode` is set. Probably the same when `isInNodeCreationDialog` is true.

File: Resources/Private/JavaScript/media-module/src/components/TopBar/SortOrderSelector.tsx [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+                value: SORT_BY.Name,
+                label: translate('sortingState.sortBy.values.name', 'Name'),
+                icon: hideOptionIcon ? '' : 'font',
+            },
+        ];
+    }, [translate]);
+
+    return (
+        <div className={classes.sortingState}>
+            <div className={classes.selectBox}>
```
Comment (suffle) [inline Resources/Private/JavaScript/media-module/src/components/TopBar/SortOrderSelector.tsx:66]: The selectbox itself doesn't seem to respect classnames

**Comment by suffle** [comment]
I tweaked the responsive behaviour a little, maybe we should think about making the Topbar multiline and collapsible on small screens?

File: Resources/Private/JavaScript/media-module/src/components/TopBar/SortOrderSelector.tsx (hunk 4 above)
Comment (Sebobo) [reply Resources/Private/JavaScript/media-module/src/components/TopBar/SortOrderSelector.tsx:66]: Hm it should. The component supports it. Maybe a JSS related issue. Will check.

Comment (Sebobo) [reply Resources/Private/JavaScript/media-module/src/components/TopBar/SortOrderSelector.tsx:66]: OK the `className` is just added to the inner div. Your solution is fine for now. To do it proper, one would have to use the `themr` api.

**Comment by Sebobo** [comment]
Ok works great. Can you remove the compiled css & js from your commits. Then I would merge and optimise the responsive behaviour in general a bit more afterwards.

**Comment by suffle** [comment]
I was wondering if the compiled files should be in there, but I guess thats just for the ci build? Removed them and fixed the tests I broke. Can you give me a little nudge, when you are done optimizing? Would love to see, what you did

**Pull request merged**